### PR TITLE
KTOR-3538 AttributeKey instance is identified by its identity instead of its name

### PR DESCRIPTION
**Pull request by lolilofit**
Ticket: [KTOR-3538](https://youtrack.jetbrains.com/issue/KTOR-3538)

**Solution**
Add `equals` and `hashCode` to the `AttributeKey` class.
Deprecate `EquatableAttributeKey`

### CODE REVIEW

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 1]
```diff
@@ -11,13 +11,30 @@ package io.ktor.util
...
+    }
 }
 
 /**
  * Version of [AttributeKey] that overrides [equals] and [hashCode] using [name]
  * @param T is type of the value stored in the attribute
  * @param name is a name of the attribute
  */
+@Suppress("DEPRECATION")
+@Deprecated("Please use `AttributeKey` class instead")
```
Comment (rsinukov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:37]: Please replace internal usages of this class

**Review by rsinukov — CHANGES_REQUESTED** [review]

**Comment by olme04** [comment]
Maybe it's better to make `EquatableAttributeKey` a deprecated typealias for `AttributeKey`?
And as we a here, I would propose to add check in `init` that name isn't blank (`require(name.isNotBlank())`) and adapt `toString` according this change. Now, as all keys are the same based on key, there is no any sense to have possibility to specify blank name

**Comment by lolilofit** [comment]
> Maybe it's better to make `EquatableAttributeKey` a deprecated typealias for `AttributeKey`? And as we a here, I would propose to add check in `init` that name isn't blank (`require(name.isNotBlank())`) and adapt `toString` according this change. Now, as all keys are the same based on key, there is no any sense to have possibility to specify blank name

@olme04 
Thanks for your comment! I've added name check in init and referenced to the `AttributeKey` in `EquatableAttributeKey` with replaceWith

**Comment by olme04** [comment]
@lolilofit and what about typealias instead of `open class` and inheriting? Now, both classes are fully identical in implementation.

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 2]
```diff
@@ -7,17 +7,43 @@ package io.ktor.util
 /**
  * Specifies a key for an attribute in [Attributes]
  * @param T is type of the value stored in the attribute
- * @param name is a name of the attribute for diagnostic purposes
+ * @param name is a name of the attribute for diagnostic purposes. Couldn't be blank
  */
 public open class AttributeKey<T>(public val name: String) {
-    override fun toString(): String = if (name.isEmpty()) super.toString() else "AttributeKey: $name"
+    init {
+        if(name.isEmpty()) {
```
Comment (rsinukov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:14]: Please fix formatting here

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 3]
```diff
@@ -7,17 +7,43 @@ package io.ktor.util
 /**
  * Specifies a key for an attribute in [Attributes]
  * @param T is type of the value stored in the attribute
- * @param name is a name of the attribute for diagnostic purposes
+ * @param name is a name of the attribute for diagnostic purposes. Couldn't be blank
  */
 public open class AttributeKey<T>(public val name: String) {
```
Comment (rsinukov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:12]: While you are here, can you please change the type of `<T>` to `<T : Any>`? `Attributes` class uses this restriction

**Comment by lolilofit** [comment]
> @lolilofit and what about typealias instead of `open class` and inheriting? Now, both classes are fully identical in implementation.

@olme04 
I've added typealias with new commits. Thanks for the help!

**Comment by lolilofit** [comment]
@andreyaksenov, please could you check public API?

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 4]
```diff
@@ -7,24 +7,22 @@ package io.ktor.util
 /**
  * Specifies a key for an attribute in [Attributes]
  * @param T is type of the value stored in the attribute
- * @param name is a name of the attribute for diagnostic purposes
+ * @param name is a name of the attribute for diagnostic purposes. Couldn't be blank
```
Comment (andreyaksenov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:10]: Couldn't be blank -> Can't be blank

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 5]
```diff
@@ -36,6 +34,18 @@ public class EquatableAttributeKey<T>(name: String) : AttributeKey<T>(name) {
     }
 }
 
+/**
+ * Version of [AttributeKey] that overrides [equals] and [hashCode] using [name]
```
Comment (andreyaksenov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:38]: Version -> A version

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 6]
```diff
@@ -36,6 +34,18 @@ public class EquatableAttributeKey<T>(name: String) : AttributeKey<T>(name) {
...
+ * @param T is type of the value stored in the attribute
```
Comment (andreyaksenov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:39]: is type -> is a type

File: ktor-utils/common/src/io/ktor/util/Attributes.kt [hunk 7]
```diff
@@ -36,6 +34,18 @@ public class EquatableAttributeKey<T>(name: String) : AttributeKey<T>(name) {
...
+ */
+@Suppress("DEPRECATION")
+@Deprecated(
+    "Please use `AttributeKey` class instead",
+    replaceWith = ReplaceWith("AttributeKey")
+)
+public typealias EquatableAttributeKey<T> = AttributeKey<T>
+
 /**
  * Create attributes instance suitable for the particular platform
```
Comment (andreyaksenov) [inline ktor-utils/common/src/io/ktor/util/Attributes.kt:50]: > Creates an attributes instance ...

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

**Review by andreyaksenov — APPROVED** [review]

**Review by rsinukov — APPROVED** [review]

**Pull request merged**